### ISSUE TITLE
Automatic download of public keys from directory servers

### ISSUE THREAD

**Issue by modir** [RainLoop/rainloop-webmail#860]
It would help a lot if PGP public keys can be automatically looked up and downloaded from e.g. https://sks-keyservers.net/ or http://keyserver.pgp.com/ or another key server.

This way people do not have to send each other they keys and if a person switches the computer he doesn't need to import all the keys again.

**Comment by kimusan** [comment]
great idea. I would love to see that implemented

**Comment by letoams** [comment]
gpg has recently added support for https://tools.ietf.org/html/draft-ietf-dane-openpgpkey-06 to do just that

**Comment by ugosan** [comment]
This is an awesome idea, I would like to work on that.

**Comment by alertme-edwin** [comment]
Now that [RFC-7929](https://tools.ietf.org/html/rfc7929) has been published, it's the perfect time to add support for automatically downloading public keys via secure DNS.  This would make transparently-applied OpenPGP encryption available even to non-experts (at least on a TOFU basis).

**Comment by ntimo** [comment]
I would like to add my +1 for this too. This would be the best feature!

**Comment by prologic** [comment]
Support for keybase.io would be nice too

**Comment by chamunks** [comment]
Yes please, of course, do not do this by default some people might not want unnecessary queries leaking out.

**Comment by Leopere** [comment]
Thanks for the reading material, I wonder what the benefits are for WKD I'm
looking forward to some time to go over it's spec.  I would worry that you
could be a method for people to query a list of possible email addresses to
spam.

**Comment by AquaL1te** [comment]
Probably is a good plan to read the docs indeed ;)

The benefits are that keys found in the WKD are automatically trusted, so you don't have to rely on manually assigning trust or building a web of trust. When the owner of a domain has a WKD, then it's assumed those keys are trusted. Because like I already stated, keys on the keyservers can be uploaded by anyone.

When you configure Apache without the indexing option, then finding keys becomes an exhaustive search. So in other words, if you don't make the WKD browsable, then privacy is pretty much there. However, there is the option to use hashes, similar to NSEC3 in DNSSEC. You can see examples when you run `gpg2 --with-wkd-hash -K`.

**Comment by letoams** [comment]
So any Certificate Agency  in the world can forge a pgp certificate then ?

I prefer RFC 7929 which puts the keys into DNSSEC. Less parties that can forge.

Paul

**Comment by AquaL1te** [comment]
Not sure what you mean, if your WKD is also using TLSA (also DANE), then it's just as secure. The benefit that WKD allows you is that the users of your domain can manage their own keys. Which lowers the administrative burden. If you have 100+ people using PGP, then it's not much fun putting every key signature in DNS.

In any case, both solutions are nice and not mutual exclusive. I'm just putting this suggestion here as well since WKD is just awesome and used in many clients/providers already.

**Comment by alertme-edwin** [comment]
Even with TLSA there is, I think, an edge case that an active attacker could exploit.  Suppose the DNSSEC response securely referenced a WKD server, but packets to that server (on port 443) are being intercepted and dropped.  A client might fall back to skipping the WKD step and possibly send the email unencrypted (or using a compromised key from a centralised server).

Also, I don't know if there is much less administrative burden in running a script which allows users to securely update paths under .well-known, compared to a script that allows them to update their key in the DNS.  Perhaps it is fair to say, though, that managing a web server is generally more easy (or more common) than managing a DNS server.

**Comment by AquaL1te** [comment]
I find these theories a bit baseless and unnecessary, Mozilla thinks it's mature and secure enough to implement like many others, including the gnupg developers themselves. Please check the official documentation, it's all documented. If you still have questions or want to discuss the design choices of WKD, then please contact the gnupg developers. There is no need to discuss it here, I only pointed out that there is WKD, that it's already adopted by software and services and that it works really nice. It's now up to the devs if they act on this feature request or not.

**Comment by letoams** [comment]
On Mon, 12 Nov 2018, Kees de Jong wrote:

> I find these theories a bit baseless and unnecessary, Mozilla thinks it's mature and secure enough to implement like many others, including the gnupg developers
> themselves.

The browsers' disdain for DNSSEC is clear. The amount of specifications
(RFC 6962, 6962bis, 8461, etc) to avoid using it is beginning to dwarf the DNSSEC
specification :P

Note that OPENPGPKEY records are supported with gnupg as well. It is
also supported with the openpgpkey tool in the hash-slinger project:

 	openpgpkey --fetch paul@nohats.ca

> Please check the official documentation, it's all documented. If you still have questions or want to discuss the design choices of WKD, then please contact
> the gnupg developers. There is no need to discuss it here, I only pointed out that there is WKD, that it's already adopted by software and services and that it works
> really nice. It's now up to the devs if they act on this feature request or not.

It is too bad the community couldnt come up with a single approach, as
fragmentation has always been pgp's weakest point. But this is not the
biggest issue of pgp not getting widespread usage. So good luck with
WKD.

Paul

**Comment by AquaL1te** [comment]
Then you also use the CAA resource record + certificate pinning. The Rainloop devs are really not waiting for a discussion about this in their bug report system. They have the info they need, now they can consider it. If you don't agree with WKD? Then that's fine, but contact the gnupg devs.
